aten/src/ATen/cuda/Atomic.cuh

```diff
@@ -35,6 +35,26 @@ struct AtomicFPOp<at::Half> {
   }
 };
 
+template <>
+struct AtomicFPOp<c10::complex<float>> {
+  template <typename func_t>
+  inline __device__ c10::complex<float> operator() (c10::complex<float> *address, c10::complex<float> val, const func_t& func) {
+    unsigned long long int* addr_as_ull = (unsigned long long int*)address;
+    unsigned long long int old = *addr_as_ull;
+    unsigned long long int assumed, new_val;
+
+    c10::complex<float> csum;
+    do {
+        assumed = old;
+        csum = func(csum, val);
+        new_val = *reinterpret_cast<unsigned long long*>(&csum);
+        old = atomicCAS(addr_as_ull, assumed, new_val);
+    } while (assumed != old);
+
+    return *reinterpret_cast<c10::complex<float>*>(&addr_as_ull);
```

Comment:
So `complex<float>` is more straight forward than `complex<double>`. I am leaving it as this implementation for now. To implement `complex<double>`, one avenue that could be taken is to use locking for example but there will be performance costs. 

However, `atomicCAS` does support 128 bits on the hopper architecture.